### ISSUE TITLE
After timeout of session, get crashed

### ISSUE THREAD

**Issue by JinGuoGe** [open62541/open62541#1071]
Description
===========
I am using source aa717f20e2. 
I am using open62541 for opcserver in Ubuntu 15.04 and in other side, using eclipse milo for opcclient in Windows.

Problem is when end of session activation after losting opcclient connection unexpectedly, then open62541 based opcserver get crashed.

Background Information / Reproduction Steps
===========================================

While monitoring of session between opcclient and opcserver, i plugged off opcserver running pc's network.
In that case after 1~2 min (i think at end of session time), then get crashed.
Please refer crashed log  here: https://paste2.org/t9LpULgU

Strange thing is sometimes doesn't happen crash.
I also tested 42187bba03	version too, and problem is occured in that case, anyway please  help fixing about unpextected disconnection  between opcserver and opcclient while keeping monitoring session.
Now there is not any valgrind errors and memory leak.

Checklist
=========
Please provide the following information:

 - [x] open62541 Version (release number or git tag): aa717f20e2
 - [x] Other OPC UA SDKs used (client or server):  eclipse milo
 - [x] Operating system: ubuntu15.04
 - [ ] Logs (with `UA_LOGLEVEL` set as low as necessary) attached
 - [ ] Wireshark network dump attached
 - [ ] Self-contained code example attached
 - [x] Critical issue

**Comment by JinGuoGe** [comment]
Hi,  thanks for trying to solve now.
I am going to add valgrind log by hoping that this can help to find reason.
https://paste2.org/LyIC0VFd

I am anyalzing problem, As result, i think reason seems that after removing jobs by removeRepeatedJob() function,  processRepeatedJobs() is called at next iteration of running loop.  (UA_Server_run_iterate() function)

**Comment by jpfr** [comment]
According to the valgrind log, the problem is sampling of values for a session that no longer exists.

First, this happens in removeRepeatedJob:
```c
    LIST_FOREACH_SAFE(rj, &server->repeatedJobs, next, rj_tmp) {
        if(!UA_Guid_equal(jobId, &rj->id))
            continue;
        LIST_REMOVE(rj, next);
        UA_free(rj);
        break;
}
```

Afterwards, there is a fault access in processRepeatedJobs:
``c
  LIST_FOREACH_SAFE(rj, &server->repeatedJobs, next, tmp_rj) {
        if(rj->nextTime > current)
            break;

         // process the job
    }
```

Something is odd about that.
You can see for yourself, that the pointer to the repeated job is no longer visible after removeRepeatedJob. So how can it be accessed?

You had some issues with multithreading before. Are you sure there is only on thread that calls UA_Server_run_iterate?

**Comment by JinGuoGe** [comment]
There is only one thread that calls UA_Server_run_iterate. This problem is occurred when requested with multiple subscriptions, but when requested with one subscription (all monitoring request are from one subscription), then this problem was not occured.

**Comment by jpfr** [comment]
Hello @JinGuoGe ,

we have to reproduce the error case to fix the issue and prevent future regressions.
For this, I added support for unit tests with simulated (precise) timings in #1089 .

If you have a dump of the network traffic (see the Wireshark tool) with the crash happening, writing the unit test can be done easily. Otherwise, it will take more time to reproduce the issue.

**Comment by jpfr** [comment]
Hello there,

I rigged a setup where the virtual machine is killing it's network driver to simulate removing a network cable. (`ifdown / ifup` on linux). I let the securechannel and session time out, activate the network driver and reconnect the client.

This runs in a loop since a day now. So far, I did not see the crash you reported.
The process has valgrind and a debugger attached. So it stops at the first error.

Every information is helpful.
- Which client did you use. How many sessions and subscriptions did you run with which settings?
- If you see the crash again, can you provide a dump of the network traffic? See here: https://github.com/open62541/open62541/wiki/Capturing-traffic-with-Wireshark

We now can simulate deterministic timings during the unit tests. So when we get a traffic dump, we can reproduce the exact same conditions.

**Comment by JinGuoGe** [comment]
Hi, @jpfr 
Sorry for delay. I 've made test program which occurred issue. 
This is test program. (https://www.dropbox.com/s/6gccyuelodykwyh/report_to_github.zip?dl=0)
I tested eclipse milo as opcua client. (https://github.com/eclipse/milo)
Same problem was occurred when using other opcua client program which based on eclipse milo. (https://github.com/comtel2000/opc-ua-client).
This is screenshot of my testing environment.
![my-environment-screenshot](https://user-images.githubusercontent.com/15504060/27992228-a2b58522-64c1-11e7-8fe7-1f03004f761f.png)
OPCUaServer is running in ubuntu, and OPCUaClient is running in windows by running java program.
I think now it is enough to generate issue in your side, because i provided test opcua server code and also opc ua client tool.
In screenshot, you can see i disabled network connection of ubuntu system.
Looking forward to your figuring problem reason.
Thanks.

**Comment by jpfr** [comment]
Thanks. I could reproduce.
Fixed on my setup via #1111.

Please reopen if the issue remains.

**Comment by JinGuoGe** [comment]
Hi, thanks.
I tested with updated version (#3702ffe7f3), but aborted by signal.
![my-environment-screenshot_2](https://user-images.githubusercontent.com/15504060/28234883-6662c24a-6937-11e7-9fcf-062735af565e.png)
This is gdb log. 
https://paste2.org/Ua8cx4jF
